# Set initialization file
https://cccma.gitlab.io/classic/makeInputFiles.html

Initialize the model for CLASSIC (CLASS+CTEM) run.

In [1]:
# Env: sc2_v0

import xarray as xr
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import proplot as pplt # New plot library (https://proplot.readthedocs.io/en/latest/)
pplt.rc['savefig.dpi'] = 300 # 1200 is too big! #https://proplot.readthedocs.io/en/latest/basics.html#Creating-figures

In [2]:
exp = 'Ref_30min_ext'
site = 'tvc'
site_ex = 'CA-Oas'

path_in = '/home/lalandmi/eccc/classic-develop/inputFiles/FLUXNETsites/'+site_ex # example file
path_out = '/home/lalandmi/eccc/classic-develop/inputFiles/SnowArctic/'+site

In [3]:
ds = xr.open_dataset(path_out+'/tvc_init_run_Ref_30min_ext_leap.nc')
ds.load()

<xarray.Dataset> Size: 11kB
Dimensions:            (tile: 1, lat: 1, lon: 1, icp1: 6, layer: 20, ic: 5,
                        icc: 12, iccp1: 13, iccp2: 14, slope: 8)
Coordinates:
  * ic                 (ic) int32 20B 1 2 3 4 5
  * icc                (icc) int32 48B 1 2 3 4 5 6 7 8 9 10 11 12
  * iccp1              (iccp1) int32 52B 1 2 3 4 5 6 7 8 9 10 11 12 13
  * iccp2              (iccp2) int32 56B 1 2 3 4 5 6 7 8 9 10 11 12 13 14
  * icp1               (icp1) int32 24B 1 2 3 4 5 6
  * slope              (slope) int32 32B 1 2 3 4 5 6 7 8
  * tile               (tile) int32 4B 1
  * layer              (layer) float64 160B 0.05 0.15 0.25 ... 18.9 41.4 58.9
  * lat                (lat) float64 8B 68.75
  * lon                (lon) float64 8B -133.5
Data variables: (12/101)
    ALBS               (tile, lat, lon) float64 8B 0.8331
    ALIC               (tile, icp1, lat, lon) float64 48B 0.0 0.29 ... 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 48B 0.0 0.05 ... 0.0 0.0
    CLAY               (tile, layer, lat, lon) float64 160B 40.0 40.0 ... 0.0
    CMAS               (tile, ic, lat, lon) float64 40B 0.0 20.0 0.0 0.0 0.0
    Cmossmas           (tile, lat, lon) float64 8B 0.0
    ...                 ...
    stemmass_s         (tile, icc, lat, lon) float64 96B 0.0 0.0 ... 0.3715
    stmhrlos           (tile, icc, lat, lon) float64 96B 0.0 0.0 0.0 ... 0.0 0.0
    tcoldm             (lat, lon) float32 4B 99.0
    twarmm             (lat, lon) float32 4B -99.0
    tymaxlai           (tile, icc, lat, lon) float64 96B 0.0 0.0 ... 1.43 0.9248
    nh4_mass           (tile, iccp1, lat, lon) float32 52B 0.0 0.0 ... 0.0 0.0
Attributes:
    title:          CLASSIC initialization file created from: /home/acrp001/m...
    creation_date:  20190523
    history:        Wed Jun 12 17:19:01 2019: ncks -4 ./rsfile.nc ./rsfile.nc
    NCO:            4.4.2
    row_bounds:     1 1

In [4]:
ds.soilcmas.sum()

<xarray.DataArray 'soilcmas' ()> Size: 8B
array(84.52871871)

## Trail Valley Creak, northwestern Canada

The site is a low-center polygon, with herb tundra and no erect vegetation

| Short name | umt |
|:-----------|:----|
| Location | 68.74617°N, 133.50171°W |
| Elevation | 85 m |
| Snow-free albedo | - |
| Simulation period | 2013 to 2019 |
| Temperature/humidity measurement height | 2 m |
| Wind measurement height | 4 m |
| Reference | Dutch et al. ([2022](https://tc.copernicus.org/articles/16/4201/2022/)) |

Southern Arctic tundra site, erect-shrub tundra, tundra dominated by low shrubs <40 cm, continuous permafrost
Elevation: 85 m

Vegetation:
- dominant plant species: Graminoids (Carex spp.), deciduous (Vaccinum uliginosum, Betula glandulosa, Salix glauca) and evergreen
shrubs (Rhododendron groenlandicum)
- rooting depth: maximum set to 40 cm
- average plant height < 40 cm
- average LAI: 0.66 ± 0.11 m2/m2
- plant cover: 92.5 % vascular plants, 31.5 % moss, 51 % lichen
- Bowen ratio:

Climate:
- mean annual precipitation: 240.6 mm (Fluxnet)
- mean annual temperature: -8.2 degrees C (Fluxnet)

Soils:
- silty clay to silt loam
- soil organic matter content:
- continuous permafrost with average thaw depth of 0.74 ± 0.05 m

Soil composition (sand/clay/silt): 18 ± 3 cm organic layer depth  
	mineral soil   35 % sand / 40 % clay (Marsh et al., [2010](https://onlinelibrary.wiley.com/doi/10.1002/hyp.7786))  
                 21 % sand / 16 % clay (Julia Boike, [AWI](https://doi.pangaea.de/10.1594/PANGAEA.962726))

Soil permeable depth: set to 9 m 

PI: Oliver Sonnentag

In [5]:
lat = 68.74617
lon = -133.50171

# If the site description is not enough we can get the input data from gridded satellite datasets
path = '/home/lalandmi/Dropbox/data/CLASSIC/'
SoilGrids250m_CLAY = xr.open_dataset(path+'/soil/SoilGrids250m_CLAY_0.05deg.nc')
SoilGrids250m_ORGM = xr.open_dataset(path+'/soil/SoilGrids250m_ORGM_0.05deg.nc')
SoilGrids250m_SAND = xr.open_dataset(path+'/soil/SoilGrids250m_SAND_0.05deg.nc')
SoilGrids250m_SDEP = xr.open_dataset(path+'/soil/SoilGrids250m_SDEP_0.05deg.nc')
ESACCI_PFT_2000 = xr.open_dataset(path+'/ESACCI-LC-L4-PFT-Map-300m-P1Y-2000-v2.0.8.nc')
ESACCI_PFT_2010 = xr.open_dataset(path+'/ESACCI-LC-L4-PFT-Map-300m-P1Y-2010-v2.0.8.nc')
CTEM_12_PFTs_ESACCI_2010 = xr.open_dataset(path+'/CTEM_12_PFTs_ESACCI_2010_minus_water_rescaled_v2_1deg.nc')
NCAR_SOCI = xr.open_dataset(path+'/soil/NCAR_SOCI_0.5deg.nc')
c4_fraction_1deg = xr.open_dataarray(path+'/vegetation/c4_fraction_1deg.nc')

### Set the layer coords to the center soil layer

In [6]:
layer_c = []

for i in range(len(ds.DELZ)):
    if i == 0:
        layer_c.append(ds.DELZ.cumsum().values[i]/2)
    else:
        layer_c.append(ds.DELZ.cumsum().values[i-1] + ds.DELZ.values[i]/2)
        
with xr.set_options(keep_attrs=True):
    ds = ds.assign_coords(layer=ds.layer*0+layer_c)

ds.layer

<xarray.DataArray 'layer' (layer: 20)> Size: 160B
array([5.00e-02, 1.50e-01, 2.50e-01, 3.50e-01, 4.50e-01, 5.50e-01, 6.50e-01,
       7.50e-01, 8.50e-01, 9.50e-01, 1.10e+00, 1.35e+00, 1.70e+00, 2.15e+00,
       2.90e+00, 4.90e+00, 8.90e+00, 1.89e+01, 4.14e+01, 5.89e+01])
Coordinates:
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9
Attributes:
    standard_name:  ground layers

In [6]:
ds.layer

<xarray.DataArray 'layer' (layer: 20)> Size: 160B
array([5.00e-02, 1.50e-01, 2.50e-01, 3.50e-01, 4.50e-01, 5.50e-01, 6.50e-01,
       7.50e-01, 8.50e-01, 9.50e-01, 1.10e+00, 1.35e+00, 1.70e+00, 2.15e+00,
       2.90e+00, 4.90e+00, 8.90e+00, 1.89e+01, 4.14e+01, 5.89e+01])
Coordinates:
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9
Attributes:
    standard_name:  ground layers

In [7]:
ds.DELZ.cumsum()

<xarray.DataArray 'DELZ' (layer: 20)> Size: 160B
array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.2,
        1.5,  1.9,  2.4,  3.4,  6.4, 11.4, 26.4, 56.4, 61.4])
Coordinates:
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9

### Check soil data
https://essd.copernicus.org/articles/13/4331/2021/essd-13-4331-2021-supplement.pdf

Use information from Figure S1 and S2 + satellite data

#### SAND

In [12]:
SoilGrids250m_SAND.sel(lat=lat, lon=lon, method='nearest').SAND.values

array([17.23465 , 18.974829, 19.929226, 20.84138 , 20.84138 , 20.84138 ,
       21.42905 , 21.42905 , 21.42905 , 21.42905 , 22.786772, 22.786772,
       22.786772, 22.786772, 22.786772, 22.786772, 22.786772, 22.786772,
       22.786772, 22.786772], dtype=float32)

In [46]:
# SAND = [95]*20 # values reported Lackner et al. (2022)
# Add a peat layer to represent the lichen (not considered into CLASSIC)
# -> better summer albedo and soil temperatures
SAND = [-2] + [21]*(20-1) # values reported Lackner et al. (2022)
# SAND = [35]*20 # values reported Lackner et al. (2022)

# Average the last layers with satellite data
# SAND[3:] = (SAND[3:] + SoilGrids250m_SAND.sel(lat=lat, lon=lon, method='nearest').SAND.values[3:])/2
SAND

[-2,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21,
 21]

#### CLAY

In [14]:
SoilGrids250m_CLAY.sel(lat=lat, lon=lon, method='nearest').CLAY.values

array([25.570757, 28.774773, 30.386208, 31.16558 , 31.16558 , 31.16558 ,
       31.289341, 31.289341, 31.289341, 31.289341, 30.764635, 30.764635,
       30.764635, 30.764635, 30.764635, 30.764635, 30.764635, 30.764635,
       30.764635, 30.764635], dtype=float32)

In [42]:
# CLAY = [40.0]*20 # values reported Lackner et al. (2022)
CLAY = [16.0]*20 # values reported Lackner et al. (2022)

# Average the last layers with satellite data
# CLAY[3:] = (CLAY[3:] + SoilGrids250m_CLAY.sel(lat=lat, lon=lon, method='nearest').CLAY.values[3:])/2
CLAY

[16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0,
 16.0]

#### ORGM

In [16]:
SoilGrids250m_ORGM.sel(lat=lat, lon=lon, method='nearest').ORGM.values

array([27.686783, 19.636015, 15.593869, 13.160918, 13.160918, 13.160918,
       12.136015, 12.136015, 12.136015, 12.136015, 10.249042, 10.249042,
       10.249042, 10.249042, 10.249042, 10.249042, 10.249042, 10.249042,
       10.249042, 10.249042], dtype=float32)

In [17]:
# Could be better constrain with Gagnon et al. (2019)
ORGM = [8.0]*20 # values reported Lackner et al. (2022)

# Average the last layers with satellite data
# ORGM[3:] = (ORGM[3:] + SoilGrids250m_ORGM.sel(lat=lat, lon=lon, method='nearest').ORGM.values[3:])/2
ORGM

[8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0,
 8.0]

In [18]:
ESACCI_PFT_2000.sel(lat=lat, lon=lon, method='nearest').load()

<xarray.Dataset> Size: 89B
Dimensions:       (time: 1, bounds: 2)
Coordinates:
    lon           float64 8B -133.5
    lat           float64 8B 68.75
  * time          (time) datetime64[ns] 8B 2000-01-01
Dimensions without coordinates: bounds
Data variables: (12/20)
    WATER         (time) int8 1B 0
    BARE          (time) int8 1B 86
    BUILT         (time) int8 1B 0
    GRASS-MAN     (time) int8 1B 0
    GRASS-NAT     (time) int8 1B 14
    SHRUBS-BD     (time) int8 1B 0
    ...            ...
    TREES-NE      (time) int8 1B 0
    LAND          (time) int8 1B 100
    WATER_OCEAN   (time) int8 1B 0
    lat_bounds    (bounds) float64 16B 68.75 68.74
    lon_bounds    (bounds) float64 16B -133.5 -133.5
    time_bounds   (time, bounds) datetime64[ns] 16B 2000-01-01 2000-12-31
Attributes: (12/40)
    title:                      ESA CCI Land Cover Project: Maps of Plant Fun...
    summary:                    This dataset contains the global plant functi...
    type:                       ESACCI-LC-L4-PFT-Map-300m-P1Y
    id:                         ESACCI-LC-L4-PFT-Map-300m-P1Y-2000-v2.0.8
    project:                    Medium Resolution Land Cover - Climate Change...
    references:                 https://maps.elie.ucl.ac.be/CCI/viewer/downlo...
    ...                         ...
    geospatial_lat_units:       degrees_north
    geospatial_lat_resolution:  002777777777778
    geospatial_lon_units:       degrees_east
    geospatial_lon_resolution:  002777777777778
    TileSize:                   2025:2025
    product_version:            2.0.8

In [19]:
ESACCI_PFT_2010.sel(lat=lat, lon=lon, method='nearest').load()

<xarray.Dataset> Size: 87B
Dimensions:       (time: 1, bounds: 2)
Coordinates:
    lon           float64 8B -133.5
    lat           float64 8B 68.75
  * time          (time) datetime64[ns] 8B 2010-01-01
Dimensions without coordinates: bounds
Data variables: (12/18)
    WATER         (time) int8 1B 0
    BARE          (time) int8 1B 86
    BUILT         (time) int8 1B 0
    GRASS-MAN     (time) int8 1B 0
    GRASS-NAT     (time) int8 1B 14
    SHRUBS-BD     (time) int8 1B 0
    ...            ...
    TREES-BE      (time) int8 1B 0
    TREES-ND      (time) int8 1B 0
    TREES-NE      (time) int8 1B 0
    lat_bounds    (bounds) float64 16B 68.75 68.74
    lon_bounds    (bounds) float64 16B -133.5 -133.5
    time_bounds   (time, bounds) datetime64[ns] 16B 2010-01-01 2010-12-31
Attributes: (12/37)
    title:                      ESA CCI Land Cover Project: Maps of Plant Fun...
    summary:                    This dataset contains the global plant functi...
    type:                       ESACCI-LC-L4-PFT-Map-300m-P1Y
    id:                         ESACCI-LC-L4-PFT-Map-300m-P1Y-2010-v2.0.8
    project:                    Climate Change Initiative - European Space Ag...
    references:                 http://www.esa-landcover-cci.org
    ...                         ...
    geospatial_lat_units:       degrees_north
    geospatial_lat_resolution:  002777777777778
    geospatial_lon_units:       degrees_east
    geospatial_lon_resolution:  002777777777778
    TileSize:                   2025:2025
    product_version:            2.0.8

In [20]:
CTEM_12_PFTs_ESACCI_2010.sel(lat=lat, lon=360+lon, method='nearest').load()

<xarray.Dataset> Size: 76B
Dimensions:  ()
Coordinates:
    lon      float64 8B 226.0
    lat      float64 8B 68.5
Data variables: (12/13)
    NE       float32 4B 0.1737
    ND       float32 4B 0.0
    BE       float32 4B 0.0
    BCD      float64 8B 0.04806
    BDD      float64 8B 0.0
    C3C      float32 4B 0.0
    ...       ...
    C3G      float32 4B 0.1525
    C4G      float32 4B 0.0
    Sedge    float32 4B 0.04829
    SBE      float32 4B 0.03125
    SBD      float32 4B 0.2193
    BARE     float32 4B 0.3269
Attributes:
    CDI:                       Climate Data Interface version 1.6.8 (http://m...
    Conventions:               CF-1.5
    cdo_openmp_thread_number:  8
    CDO:                       Climate Data Operators version 1.6.8rc2 (http:...
    created_by:                Ed Chan, CRD
    creation_date:             Wed Nov  9 19:52:31 GMT 2022
    provenance:                #1: https://climate.esa.int/en/projects/land-c...

In [21]:
# These values can be attributed with the global grid if not available
SDEP = SoilGrids250m_SDEP.sel(lat=lat, lon=lon, method='nearest').SDEP.values.item(0) # Soil permeable depth (m)
SOCI = NCAR_SOCI.sel(lat=lat, lon=lon, method='nearest').SOCI.values.item(0) # Soil color inde
c4_fraction = c4_fraction_1deg.sel(lat=lat, lon=lon, method='nearest').values.item(0) # If grass


# Those can be distributed along the soil layers if available
# CLAY = CLAY.values
# SAND = SAND.values
# ORGM = ORGM.values
# If the sum does not reach 100 %, the left over will be attributed to silt

# ! List of CLASS-level PFTs
classpfts = ['NdlTr', 'BdlTr', 'Crops', 'Grass', 'BdlSh']
FCAN = {'NdlTr': 0. , 'BdlTr': 0., 'Crops': 0., 'Grass': 0.60, 'BdlSh': 0.33} # max 1
# Set grass to 0.7 (could be 0.5 or other) to let a little bare ground (as described for this site)

# ! List of CTEM PFTs
# ! **Note: 'BdlDCoTr' should be specified before 'BdlDDrTr' due to some code in competition.
ctempfts = ['NdlEvgTr', 'NdlDcdTr', 'BdlEvgTr', 'BdlDCoTr', 'BdlDDrTr', 'CropC3', 'CropC4', 'GrassC3', 'GrassC4', 'Sedge', 'BdlEvgSh', 'BdlDCoSh']
fcancmx = {'NdlEvgTr': 0., 'NdlDcdTr': 0., 'BdlEvgTr': 0., 'BdlDCoTr': 0., 'BdlDDrTr': 0., 'CropC3': 0., 'CropC4': 0., 
           'GrassC3': 0., 'GrassC4': 0., 'Sedge': 0.60, 'BdlEvgSh': 0.05, 'BdlDCoSh': 0.28} # Max 1
# Check differences between C3 and C4 at Cdp? (TODO)

if sum(FCAN.values()) > 1: raise Exception("The sum of FCAN values needs to be lower than 1.")
if sum(fcancmx.values()) > 1: raise Exception("The sum of fcanmax values needs to be lower than 1.")

print('SDEP = ' + str(SDEP))
print('\nSOCI = ' + str(SOCI))
print('\nCLAY = ' + str(CLAY))
print('\nSAND = ' + str(SAND))
print('\nORGM = ' + str(ORGM))
print('\nGrass C4 fraction = ' + str(c4_fraction))

SDEP = 9.84000015258789

SOCI = 16.0

CLAY = [40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0, 40.0]

SAND = [-2, -2, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35]

ORGM = [8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0]

Grass C4 fraction = 0.0


### Set lat/lon

In [22]:
with xr.set_options(keep_attrs=True):
    ds = ds.assign_coords(lat=(ds.lat*0+lat))
    ds = ds.assign_coords(lon=(ds.lon*0+lon))
ds

<xarray.Dataset> Size: 11kB
Dimensions:            (tile: 1, lat: 1, lon: 1, icp1: 6, layer: 20, ic: 5,
                        icc: 12, iccp1: 13, iccp2: 14, slope: 8)
Coordinates:
  * ic                 (ic) int32 20B 1 2 3 4 5
  * icc                (icc) int32 48B 1 2 3 4 5 6 7 8 9 10 11 12
  * iccp1              (iccp1) int32 52B 1 2 3 4 5 6 7 8 9 10 11 12 13
  * iccp2              (iccp2) int32 56B 1 2 3 4 5 6 7 8 9 10 11 12 13 14
  * icp1               (icp1) int32 24B 1 2 3 4 5 6
  * slope              (slope) int32 32B 1 2 3 4 5 6 7 8
  * tile               (tile) int32 4B 1
  * layer              (layer) float64 160B 0.05 0.15 0.25 ... 18.9 41.4 58.9
  * lat                (lat) float64 8B 68.75
  * lon                (lon) float64 8B -133.5
Data variables: (12/101)
    ALBS               (tile, lat, lon) float64 8B 0.7647
    ALIC               (tile, icp1, lat, lon) float64 48B 0.0 0.29 ... 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 48B 0.0 0.05 ... 0.0 0.0
    CLAY               (tile, layer, lat, lon) float64 160B 15.16 ... 23.29
    CMAS               (tile, ic, lat, lon) float64 40B 0.0 20.0 0.0 0.0 0.0
    Cmossmas           (tile, lat, lon) float64 8B 0.0
    ...                 ...
    stemmass_s         (tile, icc, lat, lon) float64 96B 0.0 0.0 0.0 ... 0.0 0.0
    stmhrlos           (tile, icc, lat, lon) float64 96B 0.0 0.0 0.0 ... 0.0 0.0
    tcoldm             (lat, lon) float32 4B 99.0
    twarmm             (lat, lon) float32 4B -99.0
    tymaxlai           (tile, icc, lat, lon) float64 96B 0.0 0.0 0.0 ... 0.0 0.0
    nh4_mass           (tile, iccp1, lat, lon) float32 52B 0.0 0.0 ... 0.0 0.0
Attributes:
    title:          CLASSIC initialization file created from: /home/acrp001/m...
    creation_date:  20190523
    history:        Wed Jun 12 17:19:01 2019: ncks -4 ./rsfile.nc ./rsfile.nc
    NCO:            4.4.2
    row_bounds:     1 1

### Set PFTs

In [9]:
# for i, pft in enumerate(classpfts):
#     ds.FCAN[0, i, 0, 0] = FCAN[pft]
#     print(pft + ' -> ' + str(FCAN[pft]*100) + ' %')
    
ds.FCAN[0, :, 0, 0]
# 6th PFT: Bareground (need to be specified if CLASS used without CTEM)

<xarray.DataArray 'FCAN' (icp1: 6)> Size: 48B
array([0.  , 0.  , 0.  , 0.6 , 0.33, 0.  ])
Coordinates:
  * icp1     (icp1) int32 24B 1 2 3 4 5 6
    tile     int32 4B 1
    lat      float64 8B 68.75
    lon      float64 8B -133.5
Attributes:
    units:      -
    long_name:  Annual maximum fractional coverage of modelled area (read in ...

In [10]:
# for i, pft in enumerate(ctempfts):
#     ds.fcancmx[0, i, 0, 0] = fcancmx[pft]
#     print(pft + ' -> ' + str(fcancmx[pft]*100) + ' %')
    
ds.fcancmx[0, :, 0, 0]

<xarray.DataArray 'fcancmx' (icc: 12)> Size: 96B
array([0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.6 , 0.05,
       0.28])
Coordinates:
  * icc      (icc) int32 48B 1 2 3 4 5 6 7 8 9 10 11 12
    tile     int32 4B 1
    lat      float64 8B 68.75
    lon      float64 8B -133.5
Attributes:
    units:      -
    long_name:  PFT fractional coverage per grid cell

### Set soil color

In [11]:
# with xr.set_options(keep_attrs=True):
#     ds['SOCI'][0, 0, 0] = SOCI
ds.SOCI

<xarray.DataArray 'SOCI' (tile: 1, lat: 1, lon: 1)> Size: 8B
array([[[16.]]])
Coordinates:
  * tile     (tile) int32 4B 1
  * lat      (lat) float64 8B 68.75
  * lon      (lon) float64 8B -133.5
Attributes:
    units:      index
    long_name:  Soil colour index

### Set permeable depth

In [12]:
# with xr.set_options(keep_attrs=True):
#     ds['SDEP'][0, 0, 0] = SDEP
ds.SDEP

<xarray.DataArray 'SDEP' (tile: 1, lat: 1, lon: 1)> Size: 8B
array([[[9.84000015]]])
Coordinates:
  * tile     (tile) int32 4B 1
  * lat      (lat) float64 8B 68.75
  * lon      (lon) float64 8B -133.5
Attributes:
    units:      m
    long_name:  Soil permeable depth

### Check maximum level before bedrock
https://gitlab.com/jormelton/classic/-/blob/develop/src/modelStateDrivers.f90?ref_type=heads#L968

In [13]:
ds.DELZ

<xarray.DataArray 'DELZ' (layer: 20)> Size: 160B
array([ 0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.1,  0.2,
        0.3,  0.4,  0.5,  1. ,  3. ,  5. , 15. , 30. ,  5. ])
Coordinates:
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9
Attributes:
    units:      m
    long_name:  Ground layer thickness

In [15]:
i = 0
while ds.DELZ.cumsum()[i] < ds.SDEP:
    i += 1
    if i > ds.DELZ.shape[0]-1:
        print('The permable depth is greater than the model levels')
        break
        
maxlevel = i + 1 # first level of bedrock (next to the one containing SDEP)
maxlevel

17

In [16]:
ds.DELZ.cumsum()

<xarray.DataArray 'DELZ' (layer: 20)> Size: 160B
array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.2,
        1.5,  1.9,  2.4,  3.4,  6.4, 11.4, 26.4, 56.4, 61.4])
Coordinates:
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9

In [17]:
ds.DELZ.cumsum()[maxlevel]

<xarray.DataArray 'DELZ' ()> Size: 8B
array(26.4)
Coordinates:
    layer    float64 8B 18.9

In [18]:
ds.DELZ.cumsum()[:maxlevel]

<xarray.DataArray 'DELZ' (layer: 17)> Size: 136B
array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.2,
        1.5,  1.9,  2.4,  3.4,  6.4, 11.4])
Coordinates:
  * layer    (layer) float64 136B 0.05 0.15 0.25 0.35 0.45 ... 2.15 2.9 4.9 8.9

In [32]:
ds.DELZ.cumsum()[maxlevel:]

<xarray.DataArray 'DELZ' (layer: 3)> Size: 24B
array([26.4, 56.4, 61.4])
Coordinates:
  * layer    (layer) float64 24B 18.9 41.4 58.9

### Set soil contents and flags
Note: flags are only set in the SAND variable (-3 bedrock / -2 peatland)

In [47]:
with xr.set_options(keep_attrs=True):
    # Set values until bedrock
    ds.SAND[0, :maxlevel, 0, 0] = SAND[:maxlevel]
    ds.CLAY[0, :maxlevel, 0, 0] = CLAY[:maxlevel]
    # ds.ORGM[0, :maxlevel, 0, 0] = ORGM[:maxlevel]
    
    # Set bedrock values
    # ds.SAND[0, maxlevel:, 0, 0] = -3 # flag for bedrock
    # ds.CLAY[0, maxlevel:, 0, 0] = 0.
    # ds.ORGM[0, maxlevel:, 0, 0] = 0.

In [48]:
# Current values
ds.SAND[0, :, 0, 0]

<xarray.DataArray 'SAND' (layer: 20)> Size: 160B
array([-2., 21., 21., 21., 21., 21., 21., 21., 21., 21., 21., 21., 21.,
       21., 21., 21., 21., -3., -3., -3.])
Coordinates:
    tile     int32 4B 1
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9
    lat      float64 8B 68.75
    lon      float64 8B -133.5
Attributes:
    units:      %
    long_name:  Percentage sand content

In [49]:
ds.CLAY[0, :, 0, 0]

<xarray.DataArray 'CLAY' (layer: 20)> Size: 160B
array([16., 16., 16., 16., 16., 16., 16., 16., 16., 16., 16., 16., 16.,
       16., 16., 16., 16.,  0.,  0.,  0.])
Coordinates:
    tile     int32 4B 1
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9
    lat      float64 8B 68.75
    lon      float64 8B -133.5
Attributes:
    units:      %
    long_name:  Percentage clay content

In [22]:
ds.ORGM[0, :, 0, 0]

<xarray.DataArray 'ORGM' (layer: 20)> Size: 160B
array([8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       0., 0., 0.])
Coordinates:
    tile     int32 4B 1
  * layer    (layer) float64 160B 0.05 0.15 0.25 0.35 ... 8.9 18.9 41.4 58.9
    lat      float64 8B 68.75
    lon      float64 8B -133.5
Attributes:
    units:      %
    long_name:  Percentage organic matter content

In [23]:
ds

<xarray.Dataset> Size: 11kB
Dimensions:            (tile: 1, lat: 1, lon: 1, icp1: 6, layer: 20, ic: 5,
                        icc: 12, iccp1: 13, iccp2: 14, slope: 8)
Coordinates:
  * ic                 (ic) int32 20B 1 2 3 4 5
  * icc                (icc) int32 48B 1 2 3 4 5 6 7 8 9 10 11 12
  * iccp1              (iccp1) int32 52B 1 2 3 4 5 6 7 8 9 10 11 12 13
  * iccp2              (iccp2) int32 56B 1 2 3 4 5 6 7 8 9 10 11 12 13 14
  * icp1               (icp1) int32 24B 1 2 3 4 5 6
  * slope              (slope) int32 32B 1 2 3 4 5 6 7 8
  * tile               (tile) int32 4B 1
  * layer              (layer) float64 160B 0.05 0.15 0.25 ... 18.9 41.4 58.9
  * lat                (lat) float64 8B 68.75
  * lon                (lon) float64 8B -133.5
Data variables: (12/101)
    ALBS               (tile, lat, lon) float64 8B 0.8331
    ALIC               (tile, icp1, lat, lon) float64 48B 0.0 0.29 ... 0.0 0.0
    ALVC               (tile, icp1, lat, lon) float64 48B 0.0 0.05 ... 0.0 0.0
    CLAY               (tile, layer, lat, lon) float64 160B 40.0 40.0 ... 0.0
    CMAS               (tile, ic, lat, lon) float64 40B 0.0 20.0 0.0 0.0 0.0
    Cmossmas           (tile, lat, lon) float64 8B 0.0
    ...                 ...
    stemmass_s         (tile, icc, lat, lon) float64 96B 0.0 0.0 ... 0.3715
    stmhrlos           (tile, icc, lat, lon) float64 96B 0.0 0.0 0.0 ... 0.0 0.0
    tcoldm             (lat, lon) float32 4B 99.0
    twarmm             (lat, lon) float32 4B -99.0
    tymaxlai           (tile, icc, lat, lon) float64 96B 0.0 0.0 ... 1.43 0.9248
    nh4_mass           (tile, iccp1, lat, lon) float32 52B 0.0 0.0 ... 0.0 0.0
Attributes:
    title:          CLASSIC initialization file created from: /home/acrp001/m...
    creation_date:  20190523
    history:        Wed Jun 12 17:19:01 2019: ncks -4 ./rsfile.nc ./rsfile.nc
    NCO:            4.4.2
    row_bounds:     1 1

## Save to netCDF
If CLASS would be run without CTEM further variables should be set. See: https://cccma.gitlab.io/classic/basicInputs.html -> Required vegetation data


In [37]:
exp = 'Ref2_30min_ext_leap'

In [31]:
# !rm -f {path_out}/{site}_init_spinup_{exp}.nc
# !rm -f {path_out}/rsfile_spinup_{exp}.nc

In [50]:
ds.to_netcdf(path_out+'/'+site+'_init_run_'+exp+'.nc')
ds.to_netcdf(path_out+'/rsfile_run_'+exp+'.nc')

In [39]:
site

'tvc'

In [40]:
!mkdir -p /home/lalandmi/eccc/classic-develop/outputFiles/SnowArctic/{site}/run_{exp}